### PR TITLE
Enable `parthenon::par_reduce` for MD loops with Kokkos 1D Range

### PR DESCRIPTION
**Pull request by acreyes**
## PR Summary

Replaces `KOKKOS_LAMBDA`s with templated functors in `par_dispatch(LoopPatternFlatRange, ...)` that deduce the signature of the provided `function` to allow for extra arguments used in reductions.

Makes it possible to use `par_reduce(DEFAULT_LOOP_PATTERN, ..)` & `par_reduce(name, ...)` 

## PR Checklist

- [x] Code passes cpplint
- [ ] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [ ] Change is breaking (API, behavior, ...)
  - [ ] Change is *additionally* added to CHANGELOG.md in the breaking section
  - [ ] PR is marked as breaking
  - [ ] Short summary API changes at the top of the PR (plus optionally with an automated update/fix script)
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [ ] Docs build
- [ ] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

**Review by Yurlungur — APPROVED** [review]
This is a nice fix. Thanks for implementing it!

File: src/kokkos_abstraction.hpp [hunk 1]
```diff
@@ -258,6 +258,35 @@ par_dispatch(LoopPatternMDRange, const std::string &name, DevExecSpace exec_spac
...
+  void operator()(const int &idx, FArgs &&...fargs) const {
+    int k = idx / NjNi;
+    int j = (idx - k * NjNi) / Ni;
+    int i = idx - k * NjNi - j * Ni;
+    k += kl;
+    j += jl;
+    i += il;
+    function(k, j, i, std::forward<FArgs>(fargs)...);
+  }
+};
```
Comment (lroberts36) [inline src/kokkos_abstraction.hpp]: Probably not for this PR, but I wonder if we could make this work for an arbitrary dimensional index space using `IndexRange` in `utils/` with a little bit of template magic.

**Review by lroberts36 — APPROVED** [review]
LGTM, thanks for the fix!

File: src/kokkos_abstraction.hpp (hunk 1 above)
Comment (acreyes) [reply src/kokkos_abstraction.hpp]: I think it could work. This seems to work and at least gets part way there  https://godbolt.org/z/fbYP5vW6r

Comment (lroberts36) [reply src/kokkos_abstraction.hpp]: This (https://godbolt.org/z/n48951jGK) seems to work, but I didn't explicitly pull out the function signature like you did (which I think matters for references).

Comment (acreyes) [reply src/kokkos_abstraction.hpp]: yeah I remember cuda builds had given me trouble with that. This [one](https://godbolt.org/z/4W37ejM7E) uses some more template helper structs to also handle `IndexRange`s in the constructor. I imagine you could use similar things to handle all the `par_dispatch` overloads as well

File: src/kokkos_abstraction.hpp [hunk 2]
```diff
@@ -258,6 +258,35 @@ par_dispatch(LoopPatternMDRange, const std::string &name, DevExecSpace exec_spac
...
+class FlatFunctor;
+
+template <typename F, typename... Args>
+auto MakeFlatFunctor(F &function, Args... args) {
+  return FlatFunctor<F, decltype(&F::operator())>(function, std::forward<Args>(args)...);
+}
+
+template <typename Function, typename R, typename T, typename Index, typename... FArgs>
+class FlatFunctor<Function, R (T::*)(Index, Index, Index, FArgs...) const> {
+  int NjNi, Nj, Ni, kl, jl, il;
```
Comment (pgrete) [inline src/kokkos_abstraction.hpp]: `Nj` is not used, isn't it? So I suggest to remove it.
Same for the 4D version below.

**Review by pgrete — COMMENTED** [review]
Thanks for the PR!

Out of curiosity, was there a specific use and/or performance consideration that inspired these changes?

File: src/kokkos_abstraction.hpp (hunk 2 above)
Comment (acreyes) [reply src/kokkos_abstraction.hpp]: You're right, nice catch

**Comment by acreyes** [comment]
> Thanks for the PR!
> 
> Out of curiosity, was there a specific use and/or performance consideration that inspired these changes?

Not any performance consideration. I was wanting to use `par_reduce` with similar default arguments as I was with `par_for`

File: src/kokkos_abstraction.hpp (hunk 1 above)
Comment (pgrete) [reply src/kokkos_abstraction.hpp]: Saved for future reference here https://github.com/parthenon-hpc-lab/parthenon/issues/1134

**Review by pgrete — APPROVED** [review]

**Comment by pgrete** [comment]
All tests pass. I'm force merging.

**Pull request merged**